drivers/include/pulse_counter.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2017 UC Berkeley
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_pulse_counter sensor
+ * @ingroup     drivers_sensors
+ *
+ * The connection between the MCU and the PULSE_COUNTER is based on the
+ * GPIO-interface.
+ *
+ * @{
+ *
+ * @file
+ * @brief       Driver for the PULSE_COUNTER sensor
+ *
+ * @author      Hyung-Sin <hs.kim@cs.berkeley.edu>
+ */
+
+#ifndef PULSE_COUNTER_H_
+#define PULSE_COUNTER_H_
+
+#include <stdint.h>
+#include "periph/gpio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief   Parameters needed for device initialization
+ */
+typedef struct {
+    const char  *name;         /**< name of the device connected to this pin */
+    gpio_t       gpio;         /**< GPIO pin that sensor is connected to */
+    gpio_flank_t gpio_flank;   /**< GPIO flank option */ 
+} pulse_counter_params_t;
+
+/**
+  * @brief   Device descriptor for a PULSE_COUNTER device
+  * @{
+  */
+typedef struct {
+    int16_t   *pulse_count;             /**< pulse counter */
```

Comment:
I think I am missing something: this field should store that actual counter value right? So why is it a pointer? And where does it point to? Or is the actual pointer used as counter variable? Then the `int16_t` would be strange...

Reply:
There are two reasons to make this. 
1. Pointer vs. Value: SAUL read or write function "const" dev as the parameter. So, when we include pulse counter value itself in dev, we cannot change the value while executing the SAUL read function. But using pointer is fine. The pointer points the counter value defined in auto_init_pulse_counter.c.
2. I used int16_t since SAUL's output goes to "phydat_t" which uses int16_t type.

Reply:
There are two reasons to make this.

Pointer vs. Value: SAUL read or write function has "const" dev as the parameter. So, when we include pulse counter value itself in dev, we cannot change the value while executing the SAUL read function. But using pointer is fine. The pointer points the counter value defined in auto_init_pulse_counter.c.
I used int16_t since SAUL's output goes to "phydat_t" which uses int16_t type.

Reply:
I see, though I don't like it. Seems to me its quite some effort and overhead. As the device descriptor will always be in RAM, I strongly propose to change the drivers interface to use the device descriptor as non-const, and simply use an explicit cast in the SAUL read/write functions
```c
static int read_pulse_counter(const void *dev, phydat_t *res)
{
    res->val[0] = pulse_counter_read_with_reset((pulse_counter_t *)dev);
    ...
```

Reply:
If we go this way, rather than having lying function API, would not it be be better to cast the function pointer when setting the driver ?
 
```
const saul_driver_t pulse_counter_saul_driver = {
    /* read and write functions modify `dev` pointer, but we know it is in ram, so its ok */
    .read  = (saul_read_t)read_pulse_counter,
    .write = (saul_write_t)write_pulse_counter,
    .type  = SAUL_SENSE_COUNT,
};
```

How I see this case, the problem is not that `dev` is modified, as it is in ram, but that `saul_driver_t` takes read/write functions that do not modify dev.
So the API conflict should be solved when setting `saul_driver_t`.


General question, why should saul read and write functions be expected to not modify the pointer ?
It would be more in phase with https://github.com/RIOT-OS/RIOT/pull/7723/files to have read and write taking a non const pointer.

Reply:
@haukepetersen and @cladmi, good point. I changed pulse count from a pointer to a value.
I recasted "dev" from void to pulse_counter_t in the device driver, not in the SAUL interface.